modules/ocf_mirrors/files/collect-mirrors-stats

```diff
@@ -0,0 +1,168 @@
+    with filename.open('r') as f:
+        for line in f.readlines():
+            # log format is <datetime> <pid> sent <tx> bytes received <rx> bytes total size
+            if line.find('sent') > 0:
+                stats = line.split()
+                tx += int(stats[4])
+                rx += int(stats[7])
+
+    return tx, rx
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Process mirrors logs to calculate network usage '
+                                                 'and store in ocfstats')
+    parser.add_argument('-q', '--quiet', action='store_true',
+                        help='do not print stats after collecting them')
+    parser.add_argument('-n', '--dry-run', action='store_true',
+                        help='dry-run mode, do not write statistics to the database')
+    parser.add_argument('--apache', nargs='*',
+                        help='Apache log file(s) to process')
+    parser.add_argument('--rsync', nargs='*',
+                        help='rsync log files(s) to process')
+    parser.add_argument('date', nargs='?', default=date.today() - timedelta(1),
+                        help='date for use in filtering log entries')
+
+    args = parser.parse_args()
+
+    # first, figure out what projects we're getting stats for
+    sources = []
+    for project in MIRRORS_DATA_PATH.iterdir():
+        if project.is_dir() and not project.name.startswith('.'):
+            sources.append(project.name)
+
+    sources.append('other')  # catchall
+
+    # then, figure out which Apache logs we're pulling stats from
+    if args.apache:
+        apache_log_files = args.apache
+    else:
+        apache_log_files = []
+        for log_name in APACHE_LOG_FILES:
+            # logrotate rotates the log at 6am, but this script
+            # runs at midnight. So to capture 24h of data, we need
+            # to parse the rotated log as well. (the .1 file)
+            log = APACHE_LOG_PATH / log_name
+            apache_log_files.append(log)
+            apache_log_files.append(Path(str(log) + '.1'))
+
+    # and the same for rsync
+    if args.rsync:
+        rsync_log_files = args.rsync
+    else:
+        rsync_log_files = []
+        for log_file in RSYNC_LOG_PATH.iterdir():
+            # log files are named <project>.log
+            rsync_log_files.append(log_file)
+
+    # set up data structures
+    projects = {}
+    for project in sources:
+        projects[project] = {'up': 0, 'down': 0}
+
+    # pull the data itself from apache
+    for log_filename in apache_log_files:
+        projects = process_apache_log(projects, log_filename, args.date)
+
+    # ...and rsync
+    for log_filename in rsync_log_files:
+        tx, rx = process_rsync_log(log_filename)
+        projects[log_filename.stem]['up'] += rx
+        projects[log_filename.stem]['down'] += tx
+
+    # print out some stats!
+    if not args.quiet:
+        for project in projects:
+            print('{:20} {:10} {:10}'.format(
```

Comment:
Maybe we should dynamically calculate the field widths? Probably not worth it though.

Reply:
this runs in quiet mode in prod, the print display is just for debugging